### PR TITLE
Link VMSS nodes and tasks when setting up

### PR DESCRIPTION
**Pull request by ranweiler**
## Summary of the Pull Request

Add a (backwards-compatible) data field to node state update events. Use this to link nodes and tasks as soon as the tasks have been claimed. Add a new task-level `setting_up` state to encode this.

### CODE REVIEW

File: src/pytypes/onefuzztypes/models.py [hunk 1]
```diff
@@ -494,20 +494,43 @@ class WorkerDoneEvent(BaseModel):
...
+    data: Optional[SettingUpEventData]
+
+    @validator("data")
+    def check_data(cls, data: Optional[SettingUpEventData], values: Any) -> Optional[SettingUpEventData]:
+        if data:
+            state = values.get("state")
+            if state and state != NodeState.setting_up:
+                raise ValueError("data for node state update event does not match state = %s" % state)
+
+        data
```
Comment (ranweiler) [inline src/pytypes/onefuzztypes/models.py:517]: Need a `return`.

Comment (earthmehuichapan) [reply src/pytypes/onefuzztypes/models.py:517]: Fix all the state.(Erendira-_$ all needs\%._ofcrditnow.

File: src/api-service/__app__/agent_events/__init__.py [hunk 2]
```diff
@@ -41,13 +41,37 @@ def get_node_checked(machine_id: UUID) -> Node:
...
+
+            if state == NodeState.setting_up:
+                # This field will be required in the future.
+                # For now, it is optional for back compat.
+                if state_update.data:
+                    for task_id in state_update.data.tasks:
+                        task = get_task_checked(task_id)
+                        task.state = TaskState.setting_up
+
+                        # We don't yet call `on_start()` for the task.
```
Comment (bmc-msft) [inline src/api-service/__app__/agent_events/__init__.py:64]: 👍

File: src/api-service/__app__/agent_events/__init__.py [hunk 3]
```diff
@@ -41,13 +41,37 @@ def get_node_checked(machine_id: UUID) -> Node:
...
             node.state = state
             node.save()
+
+            if state == NodeState.setting_up:
+                # This field will be required in the future.
+                # For now, it is optional for back compat.
+                if state_update.data:
+                    for task_id in state_update.data.tasks:
+                        task = get_task_checked(task_id)
+                        task.state = TaskState.setting_up
```
Comment (chkeita) [inline src/api-service/__app__/agent_events/__init__.py:62]: won't this conflict with line 95 when we set the state to running in the case where we have multiple nodes running the same task

Comment (ranweiler) [reply src/api-service/__app__/agent_events/__init__.py:62]: Yes. I guess the worst problem is that we could go from `setting_up`, to `running`, and back to `setting_up`, if we have e.g. a task with `vm_count == 2`. Also, I think that problem is not unique to this new state.

Regardless, maybe we should just remove the `Task` state, but leave `setting_up` as a `NodeTaskState`.

Comment (chkeita) [reply src/api-service/__app__/agent_events/__init__.py:62]: my thoughts exactly

Comment (ranweiler) [reply src/api-service/__app__/agent_events/__init__.py]: Handled for now by only updating if we aren't `running`.

**Review by chkeita — APPROVED** [review]

**Pull request merged**